synapse/rest/client/room_batch.py

```diff
@@ -123,6 +123,19 @@ async def on_POST(
                     errcode=Codes.INVALID_PARAM,
                 )
 
+        # Make sure that the prev_event_ids exist and aren't outliers - ie, they are
+        # regular parts of the room DAG where we know the state.
+        non_outlier_prev_events = await self.store.have_events_in_timeline(
+            prev_event_ids_from_query
+        )
+        for prev_event_id in prev_event_ids_from_query:
+            if prev_event_id not in non_outlier_prev_events:
+                raise SynapseError(
+                    HTTPStatus.BAD_REQUEST,
+                    "prev_event %s does not exist, or is an outlier" % (prev_event_id,),
+                    errcode=Codes.INVALID_PARAM,
+                )
```

Comment:
This is surprising that it's necessary. Is `get_most_recent_full_state_ids_from_event_id_list` -> `state_store.get_state_ids_for_event(...)` return state for an `outlier`? It seems to look at `state_groups` which we don't expect to exist for an `outlier` but we know it can happen somehow: https://github.com/matrix-org/synapse/issues/12201#issuecomment-1081745471

I'd expect `get_most_recent_full_state_ids_from_event_id_list` to find no state for an `outlier` and the `if not state_event_ids:` check would have fired before

Reply:
> I'd expect get_most_recent_full_state_ids_from_event_id_list to find no state for an outlier and the if not state_event_ids: check would have fired before

currently, that's the case. I'm seeking to change it in https://github.com/matrix-org/synapse/pull/12191.